osc/rdma: Handle remote completion behavior

bwbarrett opened the pull request:
This patch series started to address data corruption issues due to BTLs not providing the expected remote completion.  Fixing this required two changes: 1) "accelerated" BTLs (the fast case, OFI, GNI, etc.) require REMOTE_COMPLETION or are disabled and 2) the AM RDMA compatibility interface was refactored to only use the underlying BTL RDMA functions if the BTL provides remote completion.

To implement the alternate BTL AM RDMA behavior, we use the explicit AM RDMA interface (instead of asking AM RDMA to extend the BTL), which required refactoring most of the BTL initialization code.  To use the explicit AM RDMA interface, we ended up refactoring the communication calls a bit to avoid lots of extra if statements through the communication code.

While rewriting the alternate btl initialization, the btl selection bug in https://github.com/open-mpi/ompi/issues/7830 and https://github.com/open-mpi/ompi/issues/9630 should be fixed, including the btl selection impedance mismatch.  It is no longer possible to specify alternate BTLs in osc rdma, as it now follows the active BTL list (similar to the OB1 PML).

bwbarrett commented:
Rebased on top of the recently added SM fixes, and added a priority adjustment patch now that the OSC SM component works again.

wzamazon commented on ompi/mca/osc/rdma/osc_rdma_component.c:
```diff
@@ -999,11 +990,9 @@ static int ompi_osc_rdma_query_accelerated_btls (ompi_communicator_t *comm, ompi
-        module->btls_in_use = 0;
-        module->use_memory_registration = false;
-    }
+    ompi_osc_rdma_selected_btl_insert(module, NULL, 0);
```
In previous code, the function was sometime called with module being NULL (during initialization phase, when module was not created). So that usage is gone?

wzamazon commented:
Did a quick look, and all looks good to me. Thank you!

I will checkout the branch to read again.

awlauria commented on ompi/mca/osc/rdma/osc_rdma_component.c:
```diff
@@ -395,12 +388,11 @@ static int ompi_osc_rdma_component_query (struct ompi_win_t *win, void **base, s
+        return -1;
     }
 
     return -1;
```
The logic of this function was flipped.

```suggestion
    return OMPI_SUCCESS;
```

bwbarrett commented:
Addressed Wei's feedback about the leftover comment around `query_accelerated_btls()` and a `NULL` `module` pointer, fixed the rebase error that somehow crept in (I think it was a bad manual resolution of a rebase) that caused the rdma component to never work, and fixed an error in the am-rdma wrapper code that happened in further testing with BTLs that support get() operations but not remote completion of puts.

wzamazon commented on ompi/mca/osc/rdma/osc_rdma_component.c:
```diff
@@ -1116,11 +1102,9 @@ static int ompi_osc_rdma_query_accelerated_btls (ompi_communicator_t *comm, ompi
-        module->use_memory_registration = selected_btl->btl_register_mem != NULL;
-    }
+    ompi_osc_rdma_selected_btl_insert(module, selected_btl, 0);
+    module->btls_in_use = 1;
```
probably add an `assert(module)` here?

wzamazon commented on ompi/mca/osc/rdma/osc_rdma_component.c:
```diff
@@ -578,21 +579,21 @@ static int allocate_state_shared (ompi_osc_rdma_module_t *module, void **base, s
         } else {
             for (int i = 0 ; i < module->alternate_btl_count ; ++i) {
-                module->use_cpu_atomics &= !!(module->alternate_btls[i]->btl_flags & MCA_BTL_ATOMIC_SUPPORTS_GLOB);
+                use_cpu_atomics &= !!(module->alternate_btls[i]->btl_flags & MCA_BTL_ATOMIC_SUPPORTS_GLOB);
```
There is a bug in the original implementation. I think because you are modifying this code, maybe fix it.

The bug is that for multiple nodes case, `use_cpu_atomics` has already been set to `false`. Therefore, line 589 is not-effective.  So we will need to initialize `use_cpu_atomics` to `true` before the loop.

wzamazon reviewed (approved):
Thanks! overall looks good to me.

I left two comments. One is minor. The other is about a bug in `use_cpu_atomics` in the original code, which I think can be addressed in this PR because you are working on this part of code. or we can fix it by a separate PR.

bwbarrett commented:
rebased and addressed Wei's comments.
